### ISSUE TITLE
GetIntrinsic.js uses unsafe eval

### ISSUE THREAD

**Issue by s3curitybug** [ljharb/es-abstract#117]
https://github.com/ljharb/es-abstract/blob/v1.18.0-next.1/GetIntrinsic.js#L21

This line uses Function(), which gets blocked when used with Content Security Policy unless unsafe-eval is used.

**Comment by ljharb** [comment]
It uses it safely, however, since it’s not evaluating user input and it’s not using actual eval.

There’s no other way to detect or conditionally use syntax, and this is not a security bug, so I’m not sure what’s actionable here.

**Comment by s3curitybug** [comment]
I'm sorry, I did not mean that this automatically represents a security vulnerability. However, es-abstract is a transitive dependency of many big projects, which means that many projects will stop working if they use Content Security Policy when they upgrade to version 1.18.0.

![image](https://user-images.githubusercontent.com/26045270/97516006-83aced80-1992-11eb-9f8d-2c59489d3d42.png)

I'm not familiarized with this project, I just upgraded my project and got that problem.

**Comment by ljharb** [comment]
We have always used the function-bind dependency, which uses Function as well, so this wouldn’t be new with v1.18. Have you aliased that out in your bundle so as not to run into the problem?

**Comment by s3curitybug** [comment]
I was using @babel/preset-env@7.10.4 in my project and it was working fine. Then I upgraded to @babel/preset-env@7.12.1, which uses some dependencies that end up using es-abstract@1.18.0-next.1.

Then, my project bundle started getting blocked (I use Content Security Policy). So I clicked in the console error and I got that the line blocking the script was this one: return Function('"use strict"; return (' + expressionSyntax + ').constructor;')();

So I searched it in my node_modules, and it only showed up in es-abstract -> GetIntrinsic.js. This line was not present in version 1.17.7 (which was the one used by @babel/preset-env@7.10.4): https://github.com/ljharb/es-abstract/blob/v1.17.7/GetIntrinsic.js

Probably many people is going to upgrade to @babel/preset-env@7.12.1 in the following days / weeks.

Using Function does not cause a security vulnerability atomatically, but it means that the script will be blocked in any website using Content Security Policy. For instance, if you try to open the browser console here in GitHub and execute this, it gets blocked:

expressionSyntax = 'async function* () {}';
Function('"use strict"; return (' + expressionSyntax + ').constructor;')()

![image](https://user-images.githubusercontent.com/26045270/97518085-d4264a00-1996-11eb-9b0c-1019f9e4bd63.png)

In general, using eval and Function in production environments is not a good idea.

**Comment by s3curitybug** [comment]
Would it be possible to use this instead?

```
(async function dummy() {}).constructor
(async function* dummy() {}).constructor
```

**Comment by ljharb** [comment]
No, because that's syntax, so it will cause a parse error in older browsers/engines.

There is *nothing* wrong with using `Function` (and even, with using `eval`) when it's not run on user input - that's the only place the risk lies. CSP is a useful but very blunt tool to solve that limited security problem. I believe you can configure CSP to allow eval on your own code (which includes your bundles), but disallow it on third-party code (ie, ads) - that's what I'd suggest, if possible.

**Comment by s3curitybug** [comment]
I'm sorry but I cannot agree with that. CSP is not a "blunt tool to solve a limited security problem". Javascript projects typically use hundreds of dependencies, and any of them may present an unsafe eval, a XSS, etc. (I don't mean this is the case with es-abstract). In my opinion, not using CSP in a large project nowadays is an irresponsibility. CSP is not just to disallow malicious code in ads. It's actually an effetive way to protect your web application.

Maybe es-abstract does not present any vulnerability, but if it is preventing the projects that use it from using CSP (or forcing them to allow unsafe-eval, which pretty much defeats most of the CSP purpose), it is even worse. So yes, there is something very wrong with using Function or eval.

In my case, my project performs critical operations in the front-end, so client side security is extremely important for me. Basically, you are making me choose between heavily downgrading the security of my whole project, or not using es-abstract (and thus, babel, and thus, react), which would force me to start my project again from scratch.

If you search in the internet, you can find that a lot of effort is being put into avoiding eval and Function code for production in webpack, babel, and many other projects, just to allow the usage of CSP without unsafe-eval:

https://github.com/webpack/webpack/issues/4899
https://github.com/webpack/webpack/issues/6461
https://github.com/webpack/webpack/issues/5627
https://stackoverflow.com/questions/39163748/why-is-webpack-creating-output-js-in-form-of-string-and-using-eval-function-to-d
https://stackoverflow.com/questions/50358773/webpack-babel-loader-transpiles-code-with-eval
https://stackoverflow.com/questions/59878880/remove-eval-for-production-code-through-webpack-because-of-csp-issues
https://stackoverflow.com/questions/59053103/webpack-babel-loader-produces-output-containing-eval
https://github.com/ipfs-shipyard/ipfs-companion/issues/269
...

Browsers are also putting a lot of effor into developing CSP directives.

es-abstract is used in too many projects. It is not wise forcing all of them not to use CSP (or to use it allowing unsafe-eval). I understand that compatibility with older browsers is also important (specially for a library that is used so much). But currently, the code using Function() gets inmediately executed when the script is loaded, which means that the script gets inmediately blocked by CSP. For projects that bundle everything into just one js (like my case), this represents a big problem just because of one line in one transitive dependency.

Would it be possible at least to use syntax, and if it is not supported, falling back to Function?

**Comment by ljharb** [comment]
Don't get me wrong - I like CSP and I want to enable its use.

The JS language itself does not permit *any* way to get at the value i need except using eval or Function. Syntax errors instantly end the program, they can not be recovered from.

As I said though, `function-bind` has long been a dependency of es-abstract, which uses `Function` - it seems like this is a new problem because babel has started using `es-abstract`.

If there's a way to *detect* CSP, I could consider only providing these intrinsics in non-CSP environments, which will likely not present a problem for most users of the library - do you know of one?
